aptos-move/aptos-workspace-server/src/main.rs

```diff
@@ -99,66 +115,112 @@ async fn spawn_node(test_dir: &Path) -> Result<()> {
-            indexer_grpc_port
-        ))
-        .unwrap(),
-    );
+        let api_health_checker = HealthChecker::NodeApi(
+            Url::parse(&format!("http://{}:{}", IP_LOCAL_HOST, api_port)).unwrap(),
+        );
+        api_health_checker.wait(None).await?;
 
-    api_health_checker.wait(None).await?;
-    eprintln!(
-        "Node API is ready. Endpoint: http://127.0.0.1:{}/",
-        api_port
-    );
+        eprintln!(
+            "Node API is ready. Endpoint: http://{}:{}/",
+            IP_LOCAL_HOST, api_port
+        );
 
-    indexer_grpc_health_checker.wait(None).await?;
-    eprintln!(
-        "Transaction stream is ready. Endpoint: http://127.0.0.1:{}/",
-        indexer_grpc_port
-    );
+        Ok(api_port)
+    };
+
+    let fut_indexer_grpc = async move {
+        let indexer_grpc_port = indexer_grpc_port_rx.await?;
+
+        let indexer_grpc_health_checker = HealthChecker::DataServiceGrpc(
+            Url::parse(&format!("http://{}:{}", IP_LOCAL_HOST, indexer_grpc_port)).unwrap(),
+        );
+
+        indexer_grpc_health_checker.wait(None).await?;
+        eprintln!(
+            "Transaction stream is ready. Endpoint: http://{}:{}/",
+            IP_LOCAL_HOST, indexer_grpc_port
+        );
+
+        Ok(indexer_grpc_port)
+    };
+
+    Ok((fut_api, fut_indexer_grpc))
+}
+
+/// Starts the faucet service.
+/// The port used will be returned once the service is fully up.
+async fn start_faucet(
+    test_dir: &Path,
+    fut_node_api: impl Future<Output = Result<u16, Arc<anyhow::Error>>>,
+    fut_indexer_grpc: impl Future<Output = Result<u16, Arc<anyhow::Error>>>,
+) -> Result<u16> {
+    let api_port = fut_node_api
+        .await
+        .map_err(anyhow::Error::msg)
+        .context("failed to start faucet: node api did not start successfully")?;
+
+    fut_indexer_grpc
+        .await
+        .map_err(anyhow::Error::msg)
+        .context("failed to start faucet: indexer grpc did not start successfully")?;
 
     let faucet_run_config = RunConfig::build_for_cli(
-        Url::parse(&format!(
-            "http://{}:{}",
-            node_config.api.address.ip(),
-            api_port
-        ))
-        .unwrap(),
-        "127.0.0.1".to_string(),
+        Url::parse(&format!("http://{}:{}", IP_LOCAL_HOST, api_port)).unwrap(),
+        IP_LOCAL_HOST.to_string(),
         0,
         FunderKeyEnum::KeyFile(test_dir.join("mint.key")),
         false,
         None,
     );
 
     let (faucet_port_tx, faucet_port_rx) = oneshot::channel();
-    tokio::spawn(faucet_run_config.run_and_report_port(faucet_port_tx));
+    tokio::spawn(async move {
+        if let Err(err) = faucet_run_config.run_and_report_port(faucet_port_tx).await {
+            eprintln!("Faucet exited with error {:?}", err)
+        }
+    });
 
-    let faucet_port = faucet_port_rx.await?;
+    let faucet_port = faucet_port_rx
+        .await
+        .context("failed to receive faucet port")?;
```

Comment:
also, please use `eprintln` for any errors
https://github.com/aptos-labs/aptos-core/pull/15005